README.md
# Spring Data Generator [![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.cmeza/spring-data-generator/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.cmeza/spring-data-generator)

Spring Data Generator for JPA repositories and managers.

## Features ##
* Generate in Runtime
* Generate by Plugin
* Generate repositories for JPA Entities
* Generate managers for JPA Entities
* EntityScan wrapped annotation
* Overwrite option
* Support JpaSpecificationExcecutor<?>
* Support @EmbeddedId annotation
* @Id or @EmbeddedId primitive support
* Package structure support
* Additional extends support
* Lombok support
* Jakarta support

## Dependencies ##

* **Java 11**
* **Spring data JPA**

## Generate in Runtime ##
Download the jar through Maven:

```xml
<dependency>
  <groupId>com.cmeza</groupId>
  <artifactId>spring-data-generator</artifactId>
  <version>2.0.1</version>
</dependency>
```

The simple Spring Data JPA configuration with Java-Config looks like this: 
```java
@SDGenerator(
        entityPackage = "com.acme.model",
        repositoryPackage = "com.acme.repositories",
        managerPackage = "com.acme.managers",
        repositoryPostfix = "Repository",
        managerPostfix = "Manager",
        onlyAnnotations = false,
        debug = false,
        overwrite = false,
        additionalExtends = {
                QuerydslPredicateExecutor.class
        },
        lombokAnnotations = false,
        withComments = true,
        withJpaSpecificationExecutor = true
)
@SpringBootApplication
public class AppConfig {
    
    public static void main(String[] args) {
        SpringApplication.run(WebApplication.class, args);
    }
    
}
```

| Attribute | Required | Default | Description |
|----------|:-------------:|:------:|------------|
| entityPackage |  No | [] | Entity scan package |
| repositoryPackage | No | "" | Package where the repositories will be generated |
| managerPackage | No | "" | Package where the managers will be generated | 
| repositoryPostfix | No | "Repository" | Postfix for repositories. example: Account**Repository** |
| managerPostfix | No | "Manager" | Postfix for managers. example: Account**Manager** |
| onlyAnnotations | No | false | Scan only classes annotated with @SDGenerate or @SDNoGenerate |
| debug | No | false | Enable debug log |
| overwrite | No | false | Overwrite existing files |
| additionalExtends | No | [] | Extension of additional interfaces |
| lombokAnnotations | No | false | Support Lombok annotations |
| withComments | No | true | Class comments activation |
| withJpaSpecificationExecutor | No | true | JpaSpecificationExecutor activation |

## Generate by Plugin ##
Download the jar through Maven:
```xml
<build>
	<plugins>
		<plugin>
			<groupId>com.cmeza</groupId>
			<artifactId>spring-data-generator</artifactId>
			<version>2.0.1</version>
			<configuration>
				<entity-package>
				    <param>com.acme.model</param>
				</entity-package>
				<repository-package>com.acme.repository</repository-package>
				<repository-postfix>Repository</repository-postfix>
				<manager-package>com.acme.managers</manager-package>
				<manager-postfix>Manager</manager-postfix>
				<only-annotations>false</only-annotations>
				<overwrite>false</overwrite>
				<additional-extends>
                    <param>org.springframework.data.querydsl.QuerydslPredicateExecutor</param>
                </additional-extends>
                <lombok-annotations>false</lombok-annotations>
                <with-comments>true</with-comments>
                <with-JpaSpecificationExecutor>true</with-JpaSpecificationExecutor>
			</configuration>
		</plugin>
	</plugins>
</build>
```

| Attribute | Required | Default | Description |
|----------|:-------------:|:------:|------------|
| entity-package |  Yes | [] | Entity scan package |
| repository-package | Yes | "" | Package where the repositories will be generated |
| manager-package | Yes | "" | Package where the managers will be generated | 
| repository-postfix | No | "Repository" | Postfix for repositories. example: Account**Repository** |
| manager-postfix | No | "Manager" | Postfix for managers. example: Account**Manager** |
| onlyAnnotations | No | false | Scan only classes annotated with @SDGenerate or @SDNoGenerate |
| overwrite | No | false | Overwrite existing files |
| additional-extends | No | [] | Extension of additional interfaces |
| lombok-annotations | No | false | Support Lombok annotations |
| with-comments | No | true | Class comments activation |
| with-JpaSpecificationExecutor | No | true | JpaSpecificationExecutor activation |

#### Generate repositories (terminal)
```
$ mvn spring-data-generator:repositories
```
#### Generate managers (terminal)
```
$ mvn spring-data-generator:managers
```

## Example ##

Sample entity in `com.acme.model`

```java

//import javax.persistence.Entity; //javax persistence support
//import javax.persistence.Id; //javax persistence support
import jakarta.persistence.Entity; //jakarta persistence support
import jakarta.persistence.Id; //jakarta persistence support

@Entity
//@SDGenerate   ->  Optional: Include to classes scan
//@SDNoGenerate   ->  Optional: Exclude to classes scan
public class Account {

    @Id
    @GeneratedValue
    private Long id;
    private String firstname;
    private String lastname;
       
    // Getters and setters
    // (Firstname, Lastname)-constructor and noargs-constructor
    // equals / hashcode
}
```

Generate a repository interface example in `com.acme.repositories`:

```java
// With JpaSpecificationExecutor interface
@Repository
public interface AccountRepository extends JpaRepository<Account, Long>, JpaSpecificationExecutor<Account> {
}
```

```java
// Without JpaSpecificationExecutor interface
@Repository
public interface AccountRepository extends JpaRepository<Account, Long> {
}
```

Generate a manager class example in `com.acme.managers`:

Without Lombok
```java
@Component
public class AccountManager {
    
    private final AccountRepository accountRepository;
    
    @Autowired
    public AccountManager(AccountRepository accountRepository) {
        this.accountRepository = accountRepository;
    }   
}
```
With Lombok

```java
@Component
@RequiredArgsConstructor
public class AccountManager {

    private final AccountRepository accountRepository;
    
}
```

## Notes ##

* The overwrite option delete the existing file to regenerate
* The generation of repositories and managers inherits the entity package. Example:

![alt tag](https://user-images.githubusercontent.com/9298942/39490821-bf4b53b6-4d4f-11e8-9853-fb4ece43346e.png "Package structure support")

License
----

MIT


CHANGELOG.md
# Change Log

## [2.0.1] - 2023-11-29
### Added
- [Issue:33]: Enable/Disable JpaSpecificationExecutor<?>

## [2.0.0] - 2023-02-09
### Added
- Jakarta persistence support

## [1.1.9] - 2022-04-27
### Added
- [Issue:25]: Hide class comments
- Lombok configuration
- Access "final" to manager attribute

## [1.1.8] - 2018-08-13
### Added
- [Issue:20]: Missing class while generating managers with Maven plugin

## [1.1.7] - 2018-05-30
### Added
- Package structure support

## [1.1.6] - 2018-04-09
### Added
- @Id or @EmbeddedId primitive support

## [1.1.5] - 2017-12-21
### Fix
- Fix @Id or @EmbeddedId inherited from a superclass
- Fix table info

## [1.1.4] - 2017-11-29
### Added
- Support JpaSpecificationExcecutor<?> in generated JpaRepository

## [1.1.3] - 2017-07-17
### Added
- Getter method annotated with @Id
- Support @EmbeddedId annotation

## [1.1.2] - 2017-05-23
### Optimization
- Code optimization with Sonar

## [1.1.1] - 2017-05-10
### Added
- @Repository annotation in repositories
- Overwrite option

## [1.1.0] - 2017-05-01
### Added
- Optional Scan Annotations 
- Generate by plugin

## [1.0.0] - 2017-04-09
### Added
- Generate in Runtime

[2.0.1]: https://github.com/cmeza20/spring-data-generator/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/cmeza20/spring-data-generator/compare/1.1.9...2.0.0
[1.1.9]: https://github.com/cmeza20/spring-data-generator/compare/1.1.8...1.1.9
[1.1.8]: https://github.com/cmeza20/spring-data-generator/compare/1.1.7...1.1.8
[1.1.7]: https://github.com/cmeza20/spring-data-generator/compare/1.1.6...1.1.7
[1.1.6]: https://github.com/cmeza20/spring-data-generator/compare/1.1.5...1.1.6
[1.1.5]: https://github.com/cmeza20/spring-data-generator/compare/1.1.4...1.1.5
[1.1.4]: https://github.com/cmeza20/spring-data-generator/compare/1.1.3...1.1.4
[1.1.3]: https://github.com/cmeza20/spring-data-generator/compare/1.1.2...1.1.3
[1.1.2]: https://github.com/cmeza20/spring-data-generator/compare/1.1.1...1.1.2
[1.1.1]: https://github.com/cmeza20/spring-data-generator/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/cmeza20/spring-data-generator/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/cmeza20/spring-data-generator/tree/1.0.0


[Issue:20]: https://github.com/cmeza20/spring-data-generator/issues/20
[Issue:25]: https://github.com/cmeza20/spring-data-generator/issues/25
[Issue:33]: https://github.com/cmeza20/spring-data-generator/issues/33

src/main/java/com/cmeza/sdgenerator/util/Constants.java
package com.cmeza.sdgenerator.util;

/**
 * Created by carlos on 01/05/17.
 */
public class Constants {

    public static final String ENTITY_PACKAGE = "entity-package";
    public static final String REPOSITORY_PACKAGE = "repository-package";
    public static final String REPOSITORY_POSTFIX = "repository-postfix";
    public static final String MAGANER_PACKAGE = "manager-package";
    public static final String MANAGER_POSTFIX = "manager-postfix";
    public static final String ONLY_ANNOTATIONS = "only-annotations";
    public static final String OVERWRITE = "overwrite";
    public static final String EXTENDS = "additional-extends";
    public static final String LOMBOK_ANOTATIONS = "lombok-annotations";
    public static final String WITH_COMMENTS = "with-comments";
    public static final String WITH_JPASPECIFICATIONEXECUTOR = "with-JpaSpecificationExecutor";

    public static final String PROJECT_NAME = "Spring Data Generator";

    public static final String VERSION = "(v2.0.1)";

    public static final String TABLE_POSTFIX_COLUMN = "Postfix";

    public static final String TABLE_FILE_COLUMN = "File Name";
    public static final String TABLE_RESULT_COLUMN = "Result";
    private Constants(){}
}
